lib/ueberauth/strategy/auth0/oauth.ex

```diff
@@ -114,4 +135,25 @@ defmodule Ueberauth.Strategy.Auth0.OAuth do
 
   defp get_config_value({:system, value}), do: System.get_env(value)
   defp get_config_value(value), do: value
+
+  defp compute_configs(conn, configs) do
+    case conn do
+      %Plug.Conn{} = conn when not is_nil(configs) ->
+        with module when is_atom(module) <- Keyword.get(configs, :config_from),
+             true <- function_exported?(module, :get_domain, 1),
+             true <- function_exported?(module, :get_client_id, 1),
+             true <- function_exported?(module, :get_client_secret, 1)
+        do
+          configs |> Keyword.merge([
+            domain: apply(module, :get_domain, [conn]),
+            client_id: apply(module, :get_client_id, [conn]),
+            client_secret: apply(module, :get_client_secret, [conn])
+          ])
+        else
```

Comment:
Shouldn't we raise an error if there was a `config_from` key set but it's wrongly configured ? 

Having this as a silent error might be a cause for much head-scratching 🤔 . Let's try to make this more developer friendly by guiding the use of this configuration option in case of error.

Reply:
I believe that the `config_from` configurations will be developers own responsibility. 

Currently, on the `ConfigFrom` sample module located at `test/support/`, each of the 3 functions return a string with an Auth0 credential. In my application, my `config_from` loads from `%Plug.Conn.assigns` a cached `%Account{}` that has the Auth0 credentials stored.

What I think we can do is to improve the `with` statement from `compute_configs` function so we can notify the developers that the functions couldn't be loaded. Something like this:

```elixir
defp compute_configs(conn, configs) do
  case conn do
    %Plug.Conn{} = conn when not is_nil(configs) ->
      with module when is_atom(module) <- Keyword.get(configs, :config_from),
           {:exported_get_domain, true} <- function_exported?(module, :get_domain, 1),
           {:exported_get_client_id, true} <- function_exported?(module, :get_client_id, 1),
           {:exported_get_client_secret, true} <- function_exported?(module, :get_client_secret, 1)
      do
        configs |> Keyword.merge([
          domain: apply(module, :get_domain, [conn]),
          client_id: apply(module, :get_client_id, [conn]),
          client_secret: apply(module, :get_client_secret, [conn])
        ])
      else
        {:exported_get_domain, _} ->
          {:error, :exported_get_domain_error, configs}
            
        {:exported_get_client_id, _} ->
          {:error, :exported_get_client_id_error, configs}
          
        {:exported_get_client_secret, _} ->
          {:error, :exported_get_client_secret_error, configs}

         _ -> 
           {:error, :unexpected_error, configs}
          
      end
    _ ->
      {:error, :config_from_not_found, configs}
  end
end
```

This was based on ["Elixir's with statement is fantastic"](https://dev.to/aloukissas/elixir-s-with-statement-is-fantastic-3f2g).

Reply:
I agree that the `config_from` should be the developer's responsibility. 

I was mostly thinking that we could do the following:

```elixir
defp compute_configs(conn, configs) do
    case conn do
      %Plug.Conn{} = conn when not is_nil(configs) ->
        with module when is_atom(module) <- Keyword.get(configs, :config_from),
             {:exported, true} <- {:exported, function_exported?(module, :get_domain, 1)},
             {:exported, true} <- {:exported, function_exported?(module, :get_client_id, 1)},
             {:exported, true} <- {:exported, function_exported?(module, :get_client_secret, 1)} do
          configs
          |> Keyword.merge(
            domain: apply(module, :get_domain, [conn]),
            client_id: apply(module, :get_client_id, [conn]),
            client_secret: apply(module, :get_client_secret, [conn])
          )
        else
          {:exported, false} ->
            raise

            """
            When using `:from_config`, the given module should export 3 functions:
             - `get_domain/1`
             - `get_client_id/1`
             - `get_client_secret/1`
            """

          _ ->
            {:error, :unexpected_error, configs}
        end

      _ ->
        {:error, :config_from_not_found, configs}
    end
  end
```

That way, when there is a configuration error, we suggest to the developer what is the corresponding fix/missing function. I've used a `raise` instead of an `{:error, reason}` tuple as I believe this should fail as soon as possible.

Reply:
That's much better. I've done some changes based on this.